File: apps/api/src/moldo_api/schemas/attempt.py
from typing import Literal

from pydantic import BaseModel, ConfigDict


AttemptStatus = Literal["in_progress", "submitted", "grading", "completed", "failed"]


class ApiDto(BaseModel):
    model_config = ConfigDict(extra="forbid")


class AttemptAnswerMetadata(ApiDto):
    questionId: str
    questionOrder: int
    durationSeconds: int
    audioFileName: str | None = None
    mimeType: str | None = None
    audioStorageKey: str | None = None
    audioUrl: str | None = None
    uploadedAt: str | None = None
    recordedAt: str


class ExamAttempt(ApiDto):
    id: str
    examId: str
    status: AttemptStatus
    resultId: str | None = None
    startedAt: str
    currentQuestionOrder: int
    answers: list[AttemptAnswerMetadata]
    submittedAt: str | None = None

File: packages/types/src/attempt.ts
export type AttemptStatus = "in_progress" | "submitted" | "grading" | "completed" | "failed";

export type AttemptAnswerMetadata = {
  questionId: string;
  questionOrder: number;
  durationSeconds: number;
  audioFileName?: string;
  mimeType?: string;
  audioStorageKey?: string;
  audioUrl?: string;
  uploadedAt?: string;
  recordedAt: string;
};

export type ExamAttempt = {
  id: string;
  examId: string;
  status: AttemptStatus;
  resultId?: string;
  startedAt: string;
  currentQuestionOrder: number;
  answers: AttemptAnswerMetadata[];
  submittedAt?: string;
};
